overlord/devicestate/devicemgr.go

```diff
@@ -1877,74 +1877,81 @@ func (m *DeviceManager) Systems() ([]*System, error) {
-	}
-	snapf, err := snapfile.Open(kernelSnap.Path)
-	if err != nil {
-		return nil, nil, nil, fmt.Errorf("cannot open kernel snap: %v", err)
-	}
-	kernelInfo, err := snap.ReadInfoFromSnapFile(snapf, kernelSnap.SideInfo)
-	if err != nil {
-		return nil, nil, nil, err
-	}
-	if kernelInfo.SnapType != snap.TypeKernel {
-		return nil, nil, nil, fmt.Errorf("cannot use kernel info, expected kernel but got: %v", kernelInfo.SnapType)
-	}
-
-	// gadget info needed to check encryptionSupport and gadget contraints
-	gadgetSnap := s.EssentialSnaps()[1]
-	if gadgetSnap.Path == "" {
-		return nil, nil, nil, fmt.Errorf("internal error: cannot get gadget snap path")
-	}
-	snapf, err = snapfile.Open(gadgetSnap.Path)
-	if err != nil {
-		return nil, nil, nil, fmt.Errorf("cannot open gadget snap: %v", err)
-	}
-	gadgetYaml, err := snapf.ReadFile("meta/gadget.yaml")
-	if err != nil {
-		return nil, nil, nil, fmt.Errorf("cannot read gadget.yaml: %v", err)
-	}
-	gadgetInfo, err := gadget.InfoFromGadgetYaml(gadgetYaml, sys.Model)
-	if err != nil {
-		return nil, nil, nil, fmt.Errorf("cannot parse gadget.yaml: %v", err)
+	snapInfos := make(map[snap.Type]*snap.Info)
+	seedSnaps := make(map[snap.Type]*seed.Snap)
+	var gadgetInfo *gadget.Info
+	for i, typ := range types {
+		seedSnap := s.EssentialSnaps()[i]
+		if seedSnap.Path == "" {
+			return nil, nil, nil, nil, nil, fmt.Errorf("internal error: cannot get snap path for %s", typ)
+		}
+		snapf, err := snapfile.Open(seedSnap.Path)
+		if err != nil {
+			return nil, nil, nil, nil, nil, fmt.Errorf("cannot open snap from %q: %v", seedSnap.Path, err)
+		}
+		snapInfo, err := snap.ReadInfoFromSnapFile(snapf, seedSnap.SideInfo)
+		if err != nil {
+			return nil, nil, nil, nil, nil, err
+		}
+		if snapInfo.SnapType != typ {
+			return nil, nil, nil, nil, nil, fmt.Errorf("cannot use snap info, expected %s but got %s", typ, snapInfo.SnapType)
+		}
+		switch typ {
+		case snap.TypeGadget:
+			gadgetYaml, err := snapf.ReadFile("meta/gadget.yaml")
+			if err != nil {
+				return nil, nil, nil, nil, nil, fmt.Errorf("cannot read gadget.yaml: %v", err)
+			}
+			gadgetInfo, err = gadget.InfoFromGadgetYaml(gadgetYaml, sys.Model)
+			if err != nil {
+				return nil, nil, nil, nil, nil, fmt.Errorf("cannot parse gadget.yaml: %v", err)
+			}
+		}
+		seedSnaps[typ] = seedSnap
+		snapInfos[typ] = snapInfo
 	}
 
-	encInfo, err := m.encryptionSupportInfo(sys.Model, kernelInfo, gadgetInfo)
+	encInfo, err := m.encryptionSupportInfo(sys.Model, snapInfos[snap.TypeKernel], gadgetInfo)
 	if err != nil {
-		return nil, nil, nil, err
+		return nil, nil, nil, nil, nil, err
 	}
 	opts := &gadget.ValidationConstraints{
 		EncryptedData: encInfo.StorageSafety == asserts.StorageSafetyEncrypted,
 	}
 	if err := gadget.Validate(gadgetInfo, sys.Model, opts); err != nil {
-		return nil, nil, nil, fmt.Errorf("cannot validate gadget.yaml: %v", err)
+		return nil, nil, nil, nil, nil, fmt.Errorf("cannot validate gadget.yaml: %v", err)
 	}
```

Comment:
it's a bit unclear to me if this last part belongs to the helper or not, do we need all this info about encrypytion in general? should we validate the gadget in general here before we know if we are really encrypting or not?

Reply:
I've moved validation back to `SystemAndGadgetAndEncryptionInfo`. However, I see that gadget validation was happening already in the implementation of `encryptionSupportInfo` but in a slightly different way, not sure why is that. Maybe @mvo5 can shed some light on that.